ora_migrator--0.9.2.sql

```diff
@@ -340,6 +340,91 @@ $$DECLARE
       '       uniqueness\n'
       'FROM %I.index_exp';
 
+   partition_cols_sql text := E'CREATE FOREIGN TABLE %I.partition_columns (\n'
+      '   schema          text NOT NULL,\n'
+      '   table_name      text NOT NULL,\n'
+      '   partition_name  text NOT NULL,\n'
+      '   column_name     text NOT NULL,\n'
+      '   column_position integer NOT NULL,\n'
+      '   type            text NOT NULL,\n'
+      '   position        integer NOT NULL,\n'
+      '   values          text\n'
+      ') SERVER %I OPTIONS (table ''('
+         'SELECT table_owner,\n'
+         '       part.table_name,\n'
+         '       partition_name,\n'
+         '       column_name,\n'
+         '       column_position,\n'
+         '       partitioning_type,\n'
+         '       partition_position,\n'
+         '       high_value\n'
+         'FROM dba_tab_partitions part\n'
+         '   JOIN dba_part_tables info\n'
+         '      ON part.table_owner = info.owner\n'
+         '         AND part.table_name = info.table_name\n'
+         '   JOIN dba_part_key_columns cols\n'
+         '      ON part.table_owner = cols.owner\n'
+         '         AND part.table_name = cols.name\n'
+         'WHERE table_owner NOT IN( ' || sys_schemas || E')\n'
+         '   AND object_type = ''''TABLE''''\n'
+         '   AND partitioning_type IN (''''LIST'''', ''''RANGE'''', ''''HASH'''')\n'
+      ')'', max_long ''%s'', readonly ''true'')';
```

Comment:
I didn't check, but perhaps you know:
- Does Oracle support expressions as partition keys?
- If yes, cans this query handle them?

Reply:
As far as I know, Oracle supports key columns only. Version 12.2 introduces [multiple key columns](https://oracle-base.com/articles/12c/multi-column-list-partitioning-12cr2).

Example on Oracle 21c

```sql
CREATE TABLE part1 (c1 integer, c2 varchar2(100))
  PARTITION BY LIST (SUBSTR(c2, 1, 1)) (
    PARTITION part1_a VALUES ('a'),
    PARTITION part1_b VALUES ('b'),
    PARTITION part1_default VALUES (DEFAULT)
  );
```
```text
  PARTITION BY LIST (SUBSTR(c2, 1, 1)) (
                           *
ERROR at line 2:
ORA-00907: missing right parenthesis
```

Reply:
Other example

```sql
CREATE TABLE part1 (c1 integer, c2 varchar2(100))
  PARTITION BY LIST (c1 + 1) (
    PARTITION part1_odd VALUES (1, 3, 5),
    PARTITION part1_even VALUES (2, 4, 6),
    PARTITION part1_default VALUES (DEFAULT)
  );
```
```text
  PARTITION BY LIST (c1 + 1) (
                        *
ERROR at line 2:
ORA-00907: missing right parenthesis
```

Reply:
Thanks for checking. Then we don't have to worry about that.